### PR TITLE
SessionState, NavigationMetricsListener convert to Kotlin.

### PR DESCRIPTION
**Pull request by Lebedinsky**
## Description

Migrate to Kotlin [issue](https://github.com/mapbox/navigation-sdks/issues/166)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

## Testing

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/metrics/SessionState.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.core.constants.Constants
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.services.android.navigation.v5.internal.navigation.routeprogress.MetricsRouteProgress
+import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull
+import java.util.Date
+
+internal data class SessionState @JvmOverloads constructor(
+    var secondsSinceLastReroute: Int = -1,
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/metrics/SessionState.kt]: should be this fields `val`?

Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/metrics/SessionState.kt]: No, cause of `NavigationTelemetry#line416`

**Comment by Lebedinsky** [comment]
Do not merge right now. Waiting until `NavigationTelemetry` will migrated to Kotlin.  Look at `line 409 - private void queueRerouteEvent()` and` line 423 - private FeedbackEvent queueFeedbackEven()` of `NavigationTelemetry` class

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt [hunk 2]
```diff
@@ -407,24 +399,22 @@ internal object NavigationTelemetry : NavigationMetricListener {
...
-        val eventDate = Date()
-        val rerouteEventSessionState = navigationSessionState.toBuilder()
-            .eventDate(eventDate)
-            .eventRouteProgress(metricProgress)
-            .eventLocation(metricLocation.location)
-            .secondsSinceLastReroute(getSecondsSinceLastReroute(eventDate))
-            .build()
-
+        val currentDate = Date()
+        // TODO: Need to test is copy correct
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt]: NIT remove `TODO` if tested 😛

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt [hunk 3]
```diff
@@ -158,29 +158,23 @@ internal object NavigationTelemetry : NavigationMetricListener {
...
-                }
-            )
-            navigationSessionState = navigationBuilder.build()
-            updateLastRerouteEvent(directionsRoute)
-            lastRerouteDate = Date()
-            isOffRoute = false
-        } else {
-            // Not current off-route - update the session
-            navigationSessionState = navigationBuilder.build()
+        navigationSessionState.let {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt]: Should we add a parameter `sessionState ->` so it's clear what `it` means afterwards? Would that be more readable?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt [hunk 4]
```diff
@@ -455,53 +445,44 @@ internal object NavigationTelemetry : NavigationMetricListener {
...
-        rerouteEvent.sessionState = rerouteSessionState
-
-        NavigationMetricsWrapper.rerouteEvent(
-            rerouteEvent,
-            metricProgress,
-            rerouteEvent.sessionState.eventLocation(),
-            context
-        )
+        // Update session state with locations from before / after the reroute occurred
+        rerouteSessionState.let {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt]: Same as above re: lambda to improve readability

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt [hunk 5]
```diff
@@ -455,53 +445,44 @@ internal object NavigationTelemetry : NavigationMetricListener {
...
-            metricProgress,
-            feedbackEvent.sessionState.eventLocation(),
-            feedbackEvent.description ?: "",
-            feedbackEvent.feedbackType,
-            feedbackEvent.screenshot,
-            feedbackEvent.feedbackSource,
-            context
-        )
+        // Update sessions state with locations from before / after feedback
+        feedbackSessionState.let {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationTelemetry.kt]: Same as above re: lambda to improve readability

**Pull request merged**